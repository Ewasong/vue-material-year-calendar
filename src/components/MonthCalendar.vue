<template>
<div class="c-wrapper">
  <div class="calendar"
    @mouseup="mouseUp"
    @mouseout.self="mouseUp"
    @mousein.self="mouseUp"
  >
    <div class="calendar__title">{{ month + ' 月' }}</div>
      <div
        class="calendar__body">
        <div
        v-for="(day, key) in 7"
        :key="`title${day}`"
        class="calendar__day"

      >
        {{ showDayTitle(key) }}
      </div>
      <div
        v-for="(dayObj, key) in showDays"
        class="calendar__day"
        :key="`day${key}`"
      >
        <div
          @mouseover="dragDay(arguments,dayObj)"
          @mousedown="mouseDown(dayObj)"
          class="day"
          :class="{
            'calendar__day--otherMonth': dayObj.isOtherMonth,
            'calendar--active': dayObj.active
          }"
        >
          {{ dayObj.value }}
        </div>
      </div>
    </div>
  </div>
</div>
</template>

<script>
import dayjs from 'dayjs'
export default {
  name: 'month-calendar',
  props: {
    month: {
      type: [String, Number],
      default: () => dayjs().month() + 1
    },
    year: {
      type: [String, Number],
      default: () => dayjs().year()
    },
    value: {
      type: Array,
      default: () => []
    }
  },
  data () {
    return {
      showDays: [],
      isMouseDown: false
    }
  },
  methods: {
    initCalendar () {
      if (!this.year || !this.month) return []
      const activeDay = dayjs()
        .set('year', this.year)
        .set('month', this.month - 1)
      const firstDay = activeDay.startOf('month').day()
      const lastDate = activeDay.endOf('month').date()
      const weekRow = firstDay === 6 ? 6 : 5
      const WEEK = 7
      let day = 0
      const fullCol = Array.from(Array(weekRow * WEEK).keys())
        .map(i => {
          let value = firstDay <= i
            ? day++ % lastDate + 1
            : ''
          return {
            value,
            active: false,
            isOtherMonth: firstDay > i || day > lastDate
          }
        })
      this.showDays = fullCol
    },
    showDayTitle (day) {
      const dayMapping = ['一', '二', '三', '四', '五', '六', '日']
      return dayMapping[day]
    },
    toggleDay (dayObj) {
      if (dayObj.isOtherMonth) return
      dayObj.active = !dayObj.active
    },
    dragDay (arg, dayObj) {
      if (this.isMouseDown) this.toggleDay(dayObj)
    },
    mouseDown (dayObj) {
      this.toggleDay(dayObj)
      this.isMouseDown = true
    },
    mouseUp () {
      this.isMouseDown = false
    }
  },
  watch: {
    year (val) {
      this.initCalendar()
    }
  },
  created () {
    this.initCalendar()
  }
}

</script>

<style lang="stylus" scoped>

.c-wrapper
  padding: 10px
.calendar
  background-color #fff
  min-height 295px
  text-align center
  color rgba(#353C46, .8)
  border-radius 2px
  min-width 0
  position relative
  text-decoration none
  box-shadow 0 2px 1px -1px rgba(0,0,0,.2), 0 1px 1px 0 rgba(0,0,0,.14), 0 1px 3px 0 rgba(0,0,0,.12)
  transition transform .3s ease
  &:hover
    z-index: 2
    @media (min-width: 1024px)
      transform: scale(1.15)
      box-shadow 0 7px 21px 0 rgba(0,0,0,.1)
  .calendar__title
    font-weight bold
    flex 100%
    display flex
    align-items center
    justify-content center
    border-bottom 1px solid rgba(#C4C4C4, 0.3)
    font-size 18px
    height: 50px
    margin-bottom 12px
  .calendar__body
    display flex
    flex-wrap: wrap
    justify-content flex-start
    align-content flex-start
    padding: 0px 20px
  .calendar__day
    flex 14.28%
    display flex
    justify-content center
    align-items center
    font-size 16px
    height: 31px
  .day
    font-size 14px
    cursor pointer
    user-select none
    width: 22px
    height 22px
    color: #5DB3D4
    display flex
    justify-content center
    align-items center
    &:not(.calendar__day--otherMonth):hover
      background-color rgba(#666, 0.1)
      border-radius 5px
    &.calendar--active
      background-color: rgba(#FFBABA, .5)
      border-radius 3px
      color #BCBCBC
      position relative
    &.calendar--active:after
      content ''
      display block
      height: 10px
      width 10px
      background-color #fff
      position absolute
      top -5px
      right -5px
      border-radius 50%
      background-image url('../../public/baseline-remove_circle-24px.svg')
      background-size 100% 100%
  & .calendar__day--otherMonth
    color: #eaeaea
    cursor: auto
</style>
