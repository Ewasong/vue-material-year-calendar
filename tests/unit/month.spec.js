import { mount } from '@vue/test-utils'
import MonthCalendar from '@/components/MonthCalendar'
describe('MonthCalendar', () => {
  const wrapper = mount(MonthCalendar, {
    propsData: {
      month: 1,
      lang: 'tw',
      year: 2019
    }
  })
  it('renders title', () => {
    expect(wrapper.html()).toContain('<div class="calendar__title">一月</div>')
  })

  it('renders array of strings correctly', () => {
    wrapper.setProps({
      activeDates: ['2019-01-01', '2019-01-20']
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders array of objects correctly', () => {
    wrapper.setProps({
      activeDates: [{ date: '2019-01-13', className: '' },
        { date: '2019-01-14', className: 'info' },
        { date: '2019-01-15', className: 'warning' }]
    })
    expect(wrapper.element).toMatchSnapshot()
  })
  
})
