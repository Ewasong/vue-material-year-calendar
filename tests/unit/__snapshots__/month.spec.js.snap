// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthCalendar renders array of objects correctly 1`] = `
<div
  class="c-wrapper"
>
  <div
    class="calendar"
  >
    <div
      class="calendar__title"
    >
      一月
    </div>
     
    <div
      class="calendar__body"
    >
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        一
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        二
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        三
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        四
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        五
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        六
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        日
      </div>
       
      <div
        class="calendar__day"
      >
        <div
          class="day calendar__day--otherMonth"
        >
          
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          1
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          2
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          3
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          4
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          5
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          6
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          7
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          8
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          9
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          10
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          11
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          12
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar--active "
        >
          13
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar--active info"
        >
          14
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar--active warning"
        >
          15
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          16
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          17
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          18
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          19
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          20
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          21
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          22
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          23
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          24
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          25
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          26
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          27
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          28
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          29
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          30
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          31
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar__day--otherMonth"
        >
          1
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar__day--otherMonth"
        >
          2
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar__day--otherMonth"
        >
          3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MonthCalendar renders array of strings correctly 1`] = `
<div
  class="c-wrapper"
>
  <div
    class="calendar"
  >
    <div
      class="calendar__title"
    >
      一月
    </div>
     
    <div
      class="calendar__body"
    >
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        一
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        二
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        三
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        四
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        五
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        六
      </div>
      <div
        class="calendar__day day__weektitle"
        style="font-size: 16px;"
      >
        日
      </div>
       
      <div
        class="calendar__day"
      >
        <div
          class="day calendar__day--otherMonth"
        >
          
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar--active "
        >
          1
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          2
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          3
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          4
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          5
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          6
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          7
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          8
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          9
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          10
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          11
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          12
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          13
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          14
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          15
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          16
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          17
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          18
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          19
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar--active "
        >
          20
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          21
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          22
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          23
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          24
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          25
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          26
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          27
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          28
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          29
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          30
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day"
        >
          31
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar__day--otherMonth"
        >
          1
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar__day--otherMonth"
        >
          2
        </div>
      </div>
      <div
        class="calendar__day"
      >
        <div
          class="day calendar__day--otherMonth"
        >
          3
        </div>
      </div>
    </div>
  </div>
</div>
`;
